webrtc/browser-to-browser.md

```diff
@@ -0,0 +1,119 @@
+# WebRTC browser-to-browser
+
+| Lifecycle Stage | Maturity      | Status | Latest Revision |
+|-----------------|---------------|--------|-----------------|
+| 1A              | Working Draft | Active | r0, 2022-12-15  |
+
+## Motivation
+
+**Hole punching in the browser**
+
+Browser _A_ wants to connect to Browser node _B_ with the help of server node _R_.
+Both _A_ and _B_ can not listen for incoming connections due to the restriction of the browser platform and being behind a NAT and/or firewall.
+Note that _A_ and/or _B_ may as well be non-browser nodes behind NATs and/or firewalls.
+However, for two non-browser nodes using TCP or QUIC hole punching with [DCUtR] will be the more efficient way to establish a direct connection.
+
+On a historical note, this specification replaces the existing [libp2p WebRTC star](https://github.com/libp2p/js-libp2p-webrtc-star) and [libp2p WebRTC direct](https://github.com/libp2p/js-libp2p-webrtc-direct) protocols.
+
+## Connection Establishment
+
+1. _B_ advertises support for the WebRTC browser-to-browser protocol by appending `/webrtc-w3c` to its relayed multiaddr e.g. `/ip6/fe80::883:a581:fff1:833/udp/4001/quic/webtransport/certhash/<hash1>/p2p/<relay-peer-id>/p2p-circuit/webrtc-w3c/p2p/<b-peer-id>`.
```

Comment:
Continuing discussion here, since the other are resolved

It seems that @MarcoPolo [proposal](https://github.com/libp2p/specs/pull/497#discussion_r1113489770) brings the benefits of my [proposal](https://github.com/libp2p/specs/pull/497#discussion_r1091804742) while avoiding major changes to our current dialing schemes.

The idea that "a multi address is all you need" is nice, but when is that really necessary? Most of the time, that is already not true (the `p2p` part is added / removed depending on where you are)

Sure, it's useful for an utility like `libp2p-ping /ip4/../tcp/62268/p2p/relaypeerid/p2p-circuit/targetpeerid`, but that utility can easily optimistically connect via the relay and hope for the best. That will be a bit wasteful, since sometimes, the target peer won't have a compatible transport, but realistically very rare, since the amount of connections to "manually entered MAs" vs "automatically discovered peer MA list" is very low. And even lower for a circuit relay MA, as those are not stable (won't be in bootstrap list, for instance)

It is a bit ugly to put a "capability" in a MA, hence my original proposal, but this solution seem easy to integrate, and I find it a lot cleaner than appending to the circuit address, for reasons already explained. We can also add TCP & QUIC support in the same way

Reply:
We synced up about this, and I was wrong. I know believe webrtc-direct is a capability, not an address in itself. In the ideal world we would be able to share addresses and capabilities together (like a signed peer record). `webrtc-direct` and `dcutr` are both capabilities. The addresses come later, after the STUN and candidate exchange.

Therefore it doesn't make sense to have a multiaddr that is just `/webrtc-direct`, because that's not an address and doesn't tell us how to reach the other node. If we go out to the streets of the internet and ask how to get to "webrtc-direct" we will get nothing but blank stares.

That said, we still want to advertise this capability in a way that doesn't require a node to connect to us first. In that case I think it's okay (but a hack) to append this to an existing address. Yes, this requires effectively duplicating our addresses, but I think that's okay in the short term.

Another point is that many applications assume a single multiaddr is enough to connect to another peer. We shouldn't break existing applications.

tl;dr

In favor of appending `/webrtc-direct` to an exising multiaddr.

Reply:
Well, if I'm the only one left opposing this, I won't block progress on this. If you do change your mind, please reach me at one of my addresses:
```
/ip4/140.82.121.4/tcp/4440/p2p/Qmc3PxhMhQja8N4t7mRDyGm2vHkvcxe5Kabp2iAig1DXHb/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N
/ip6/2001:db8:3632:4284:1845:6216:7187:9018/tcp/4440/p2p/Qmc3PxhMhQja8N4t7mRDyGm2vHkvcxe5Kabp2iAig1DXHb/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N
/ip4/140.82.121.4/udp/4441/quic/webtransport/certhash/b2uaraocy6yrdblb4sfptaddgimjmmpy/p2p/Qmc3PxhMhQja8N4t7mRDyGm2vHkvcxe5Kabp2iAig1DXHb/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N
/ip6/2001:db8:3632:4284:1845:6216:7187:9018/udp/4441/quic/webtransport/certhash/b2uaraocy6yrdblb4sfptaddgimjmmpy/p2p/Qmc3PxhMhQja8N4t7mRDyGm2vHkvcxe5Kabp2iAig1DXHb/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N
/ip4/140.82.121.4/udp/4442/quic/p2p/Qmc3PxhMhQja8N4t7mRDyGm2vHkvcxe5Kabp2iAig1DXHb/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N
/ip6/2001:db8:3632:4284:1845:6216:7187:9018/udp/4442/quic/p2p/Qmc3PxhMhQja8N4t7mRDyGm2vHkvcxe5Kabp2iAig1DXHb/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N
/ip4/140.82.121.4/udp/4443/webrtc/certhash/b2uaraocy6yrdblb4sfptaddgimjmmpy/p2p/Qmc3PxhMhQja8N4t7mRDyGm2vHkvcxe5Kabp2iAig1DXHb/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N
/ip6/2001:db8:3632:4284:1845:6216:7187:9018/udp/4443/webrtc/certhash/b2uaraocy6yrdblb4sfptaddgimjmmpy/p2p/Qmc3PxhMhQja8N4t7mRDyGm2vHkvcxe5Kabp2iAig1DXHb/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N
/dns/verylongaddress.com/tcp/4440/wss/p2p/Qmc3PxhMhQja8N4t7mRDyGm2vHkvcxe5Kabp2iAig1DXHb/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N
/ip4/185.93.3.244/tcp/4440/p2p/Qmcfz2MaPjw44RfVpHKFgXwhW3uFBRBxByVEkgPhefKCJW/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N
/ip6/2ff1:ae8:1845:6216:7187:3632:4284:1890/tcp/4440/p2p/Qmcfz2MaPjw44RfVpHKFgXwhW3uFBRBxByVEkgPhefKCJW/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N
/ip4/185.93.3.244/udp/4441/quic/webtransport/certhash/b2uaraocy6yrdblb4sfptaddgimjmmpy/p2p/Qmcfz2MaPjw44RfVpHKFgXwhW3uFBRBxByVEkgPhefKCJW/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N
/ip6/2ff1:ae8:1845:6216:7187:3632:4284:1890/udp/4441/quic/webtransport/certhash/b2uaraocy6yrdblb4sfptaddgimjmmpy/p2p/Qmcfz2MaPjw44RfVpHKFgXwhW3uFBRBxByVEkgPhefKCJW/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N
/ip4/185.93.3.244/udp/4442/quic/p2p/Qmcfz2MaPjw44RfVpHKFgXwhW3uFBRBxByVEkgPhefKCJW/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N
/ip6/2ff1:ae8:1845:6216:7187:3632:4284:1890/udp/4442/quic/p2p/Qmcfz2MaPjw44RfVpHKFgXwhW3uFBRBxByVEkgPhefKCJW/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N
/ip4/185.93.3.244/udp/4443/webrtc/certhash/b2uaraocy6yrdblb4sfptaddgimjmmpy/p2p/Qmcfz2MaPjw44RfVpHKFgXwhW3uFBRBxByVEkgPhefKCJW/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N
/ip6/2ff1:ae8:1845:6216:7187:3632:4284:1890/udp/4443/webrtc/certhash/b2uaraocy6yrdblb4sfptaddgimjmmpy/p2p/Qmcfz2MaPjw44RfVpHKFgXwhW3uFBRBxByVEkgPhefKCJW/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N
/dns/otherlongaddress.com/tcp/4440/wss/p2p/Qmcfz2MaPjw44RfVpHKFgXwhW3uFBRBxByVEkgPhefKCJW/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N
/ip4/140.82.121.4/tcp/4440/p2p/Qmc3PxhMhQja8N4t7mRDyGm2vHkvcxe5Kabp2iAig1DXHb/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N/webrtc-w3c
/ip6/2001:db8:3632:4284:1845:6216:7187:9018/tcp/4440/p2p/Qmc3PxhMhQja8N4t7mRDyGm2vHkvcxe5Kabp2iAig1DXHb/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N/webrtc-w3c
/ip4/140.82.121.4/udp/4441/quic/webtransport/certhash/b2uaraocy6yrdblb4sfptaddgimjmmpy/p2p/Qmc3PxhMhQja8N4t7mRDyGm2vHkvcxe5Kabp2iAig1DXHb/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N/webrtc-w3c
/ip6/2001:db8:3632:4284:1845:6216:7187:9018/udp/4441/quic/webtransport/certhash/b2uaraocy6yrdblb4sfptaddgimjmmpy/p2p/Qmc3PxhMhQja8N4t7mRDyGm2vHkvcxe5Kabp2iAig1DXHb/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N/webrtc-w3c
/ip4/140.82.121.4/udp/4442/quic/p2p/Qmc3PxhMhQja8N4t7mRDyGm2vHkvcxe5Kabp2iAig1DXHb/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N/webrtc-w3c
/ip6/2001:db8:3632:4284:1845:6216:7187:9018/udp/4442/quic/p2p/Qmc3PxhMhQja8N4t7mRDyGm2vHkvcxe5Kabp2iAig1DXHb/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N/webrtc-w3c
/ip4/140.82.121.4/udp/4443/webrtc/certhash/b2uaraocy6yrdblb4sfptaddgimjmmpy/p2p/Qmc3PxhMhQja8N4t7mRDyGm2vHkvcxe5Kabp2iAig1DXHb/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N/webrtc-w3c
/ip6/2001:db8:3632:4284:1845:6216:7187:9018/udp/4443/webrtc/certhash/b2uaraocy6yrdblb4sfptaddgimjmmpy/p2p/Qmc3PxhMhQja8N4t7mRDyGm2vHkvcxe5Kabp2iAig1DXHb/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N/webrtc-w3c
/dns/verylongaddress.com/tcp/4440/wss/p2p/Qmc3PxhMhQja8N4t7mRDyGm2vHkvcxe5Kabp2iAig1DXHb/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N/webrtc-w3c
/ip4/185.93.3.244/tcp/4440/p2p/Qmcfz2MaPjw44RfVpHKFgXwhW3uFBRBxByVEkgPhefKCJW/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N/webrtc-w3c
/ip6/2ff1:ae8:1845:6216:7187:3632:4284:1890/tcp/4440/p2p/Qmcfz2MaPjw44RfVpHKFgXwhW3uFBRBxByVEkgPhefKCJW/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N/webrtc-w3c
/ip4/185.93.3.244/udp/4441/quic/webtransport/certhash/b2uaraocy6yrdblb4sfptaddgimjmmpy/p2p/Qmcfz2MaPjw44RfVpHKFgXwhW3uFBRBxByVEkgPhefKCJW/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N/webrtc-w3c
/ip6/2ff1:ae8:1845:6216:7187:3632:4284:1890/udp/4441/quic/webtransport/certhash/b2uaraocy6yrdblb4sfptaddgimjmmpy/p2p/Qmcfz2MaPjw44RfVpHKFgXwhW3uFBRBxByVEkgPhefKCJW/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N/webrtc-w3c
/ip4/185.93.3.244/udp/4442/quic/p2p/Qmcfz2MaPjw44RfVpHKFgXwhW3uFBRBxByVEkgPhefKCJW/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N/webrtc-w3c
/ip6/2ff1:ae8:1845:6216:7187:3632:4284:1890/udp/4442/quic/p2p/Qmcfz2MaPjw44RfVpHKFgXwhW3uFBRBxByVEkgPhefKCJW/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N/webrtc-w3c
/ip4/185.93.3.244/udp/4443/webrtc/certhash/b2uaraocy6yrdblb4sfptaddgimjmmpy/p2p/Qmcfz2MaPjw44RfVpHKFgXwhW3uFBRBxByVEkgPhefKCJW/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N/webrtc-w3c
/ip6/2ff1:ae8:1845:6216:7187:3632:4284:1890/udp/4443/webrtc/certhash/b2uaraocy6yrdblb4sfptaddgimjmmpy/p2p/Qmcfz2MaPjw44RfVpHKFgXwhW3uFBRBxByVEkgPhefKCJW/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N/webrtc-w3c
/dns/otherlongaddress.com/tcp/4440/wss/p2p/Qmcfz2MaPjw44RfVpHKFgXwhW3uFBRBxByVEkgPhefKCJW/p2p-circuit/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N/webrtc-w3c
```

And I must apologize, with this information, you have no idea if you will be able to talk to me, as I may use a different multistream-select, muxer, or security channel. Unfortunately, adding these capabilities to the multiaddresses would make the total set >80k of memory, which doesn't fit in my embedded device
Also, I won't be able to send you this set over identify, since it already blows the [proposed limit](https://github.com/libp2p/specs/pull/492), without even including an SPR

Reply:
> if I'm the only one left opposing this, I won't block progress on this.

:+1: 

Note @Menduist that the decision to put the webrtc-w3c capability into a multiaddr and the decision to append that capability to an existing relayed address instead of a new multiaddr does not mean that folks are opposed to your proposal of extended peer records, e.g. containing capabilities.

I am in favor of the exploration for extended peer records. I think they are the right solution long term.

That said, I don't think we should block this effort (webrtc-w3c) on the extended peer records. Once in place, we can advertise support for webrtc-w3c via the capability feature in extended peer records. In the meantime, I think we should use multiaddr.

I am not going to drive extended peer records in the near future. I am happy to support someone (e.g. you) in that effort though.

Reply:
@Menduist I get that advertising the `/webrtc-w3c` capability on _every_ address leads to an explosion in addresses. However, why would we do that?

When connecting through a relay, the physical transport, muxer and security protocol of `B <-> R` does not matter for `R <-> A`. Thus, advertising `/webrtc-w3c` on a single relayed address should be enough, right?

It will grow linearly with the number of relays you want to use but that is already the case today, isn't it?

The outbound side of the relayed connect starts the signaling protocol. Thus we could say that a node doesn't need to publish a relayed listen address with and without `/webrtc-w3c` but instead say that the dialing node MAY ignore the `/webrtc-w3c` bit and use the relayed connection for whatever they want.

This is kind of implied by the spec anyway because the listening node doesn't know how the dialing node found them. For all they know, they could have guessed the address. Thus, if we allow this "partwise" interpretation of a multiaddress there is no need publish multiple combinations.

Reply:
> I get that advertising the /webrtc-w3c capability on every address leads to an explosion in addresses. However, why would we do that?

My understanding of the spec / discussions is that we will do that, to keep the "a multiaddress is all you need" logic
Appending `webrtc-w3c` to a random address seem as weird as adding a standalone `/webrtc-w3c` address, IMO

Reply:
> > I get that advertising the /webrtc-w3c capability on every address leads to an explosion in addresses. However, why would we do that?
> 
> My understanding of the spec / discussions is that we will do that, to keep the "a multiaddress is all you need" logic

I don't see a requirement in here that it needs to be appended to every address, just that this is how we discover the capability.

> Appending `webrtc-w3c` to a random address seem as weird as adding a standalone `/webrtc-w3c` address, IMO

It is not a random one though. It is a kind of promise that we will offer this capability on a connection from this address. See https://github.com/libp2p/specs/issues/525 which is kind of related. Not every protocol may be offered on every connection.